arcgis-ios-sdk-samples/Maps/Show location history/LocationHistoryViewController.swift

```diff
@@ -82,67 +122,85 @@ class LocationHistoryViewController: UIViewController {
+        let simulatedDataSource = AGSSimulatedLocationDataSource()
+        do {
+            let jsonObject = try JSONSerialization.jsonObject(with: polylineJSONString.data(using: .utf8)!)
+            let routePolyline = try AGSPolyline.fromJSON(jsonObject) as! AGSPolyline
+            let densifiedRoute = AGSGeometryEngine.geodeticDensifyGeometry(
+                routePolyline,
+                maxSegmentLength: 50.0,
+                lengthUnit: .meters(),
+                curveType: .geodesic
+            )
+            simulatedDataSource.setLocationsWith(densifiedRoute as! AGSPolyline)
+        } catch {
+            presentAlert(error: error)
+        }
+        return simulatedDataSource
+    }
+    
+    private func makeOverlay(with symbol: AGSSymbol) -> AGSGraphicsOverlay {
+        let overlay = AGSGraphicsOverlay()
+        overlay.renderer = AGSSimpleRenderer(symbol: symbol)
+        return overlay
+    }
+    
     private func handleLocationStatusChange() {
         if isTracking {
             startProcessingLocationChanges()
         } else {
             stopProcessingLocationChanges()
         }
-
         updateView()
     }
     
     private func processLocationUpdate() {
         guard isTracking, let position = mapView.locationDisplay.mapLocation, position.x != 0, position.y != 0 else { return }
-        
-        let locationGraphic = AGSGraphic(geometry: position, symbol: nil)
-        locationsOverlay.graphics.add(locationGraphic)
-
-        guard let trackBuilder = trackBuilder else { return }
-
-        trackBuilder.add(position)
-        let trackGraphic = AGSGraphic(geometry: trackBuilder.toGeometry(), symbol: nil)
-        trackOverlay.graphics.add(trackGraphic)
+        pointBuilder?.points.add(position)
+        locationGraphic.geometry = pointBuilder?.toGeometry()
+        trackBuilder?.add(position)
+        trackGraphic.geometry = trackBuilder?.toGeometry()
     }
     
     private func setupLocationDisplay(_ locationDisplay: AGSLocationDisplay) {
         locationDisplay.autoPanMode = .recenter
         locationDisplay.initialZoomScale = Constants.initialZoomScale
+        // Comment out the line below to use device live location.
+        locationDisplay.dataSource = makeDataSource(from: polylineJSONString)
         
-        locationDisplay.start { (error) in
+        locationDisplay.start { [weak self] (error) in
             if let error = error {
-                print("Error starting location display: \(error.localizedDescription)")
+                self?.presentAlert(error: error)
             }
         }
     }
     
     private func setupMapView() {
         let map = AGSMap(basemap: .lightGrayCanvasVector())
         mapView.map = map
-
         map.load { [weak self, unowned map] (_) in
+            self?.pointBuilder = AGSMultipointBuilder(spatialReference: map.spatialReference)
             self?.trackBuilder = AGSPolylineBuilder(spatialReference: map.spatialReference)
```

Comment:
Repeated use of `weak` variable:

```suggestion
            guard let self = self else { return }
            let spatialReference = map.spatialReference
            self.pointBuilder = AGSMultipointBuilder(spatialReference: spatialReference)
            self.trackBuilder = AGSPolylineBuilder(spatialReference: spatialReference)
```